### ISSUE TITLE
TypeScript 2 beta: undefined lookup on object gives error

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#9684]
Given:

``` tsc
window[undefined]
```

I get:

```
main.ts(21,1): error TS2342: An index expression argument must be of type 'string', 'number', 'symbol', or 'any'.
```

I don't understand what is wrong about the expression `y[undefined]`?

**Comment by nippur72** [comment]
Also, another strange thing is:
```ts
var x; 
window[x] = "a";
```
>Type 'string' is not assignable to type 'Window'.

which is a misleading error description

**Comment by mhegazy** [comment]
@nippur72's example is because Window has an index signature on it. so this seems right to me.

**Comment by mhegazy** [comment]
For the undefined indexer, this a breaking change, to avoid that, we should allow undefined and null as index expression type if we are not in strict null mode.

**Comment by OliverJAsh** [comment]
I should have said this only occurs when strict null checking is enabled,
so it's not necessarily a breaking change

**Comment by mhegazy** [comment]
the error when using `--strictNullChecks` is expected. the one that is not is when `--strictNullChecks` is not on.

**Comment by mhegazy** [comment]
I should add that the same error is displayed in both cases.

**Comment by OliverJAsh** [comment]
Oh, I thought I only saw it when I had enabled strict null checking

**Comment by OliverJAsh** [comment]
The following only happens when strict null checking is enabled:

``` ts
const a = 1 as number | undefined;
window[a]; // error
```

**Comment by sandersn** [comment]
Fix is available for review at #10009.
